text/0000-builders.md

```diff
@@ -0,0 +1,101 @@
+- Start Date: 2020-06-25
+- RFC PR: 
+- Svelte Issue: ]
+
+# Sapper Builders
+
+## Summary
+
+This RFC proposes a method of modularising build output from Sapper for ease of deployment across a variety of targets.
+
+## Motivation
+
+Sapper's current build output comprises of a series of files:
+
+1. A build manifest
+1. A client bundle
+1. A server bundle (in non-exported mode)
+1. An (optional) legacy client bundle
+1. Static files
+
+Deploying this is relatively simple, but certainly not optimised for certain deployment strategies such as serverless (lambda), and often relies on third-party, platform-specific builders.
+
+If Sapper were to produce its output in a more modular fashion, it would facilitate the creation of a number of officially supported "builders" for major platforms, as well as opening the doors for third parties to easily produce builders which targetted other platforms.
+
+Additionally, it would ease the creation of some desirable features:
+
+1. Per-route bundle splitting
+1. [Differential rendering on a per-route basis](https://github.com/sveltejs/sapper/issues/1324) - Allowing SSR, SSG, SPA, [JAM](https://github.com/sveltejs/sapper/issues/1093), on a per-route basis.
+1. [Faster cold-boot time for serverless functions](https://github.com/sveltejs/sapper/issues/356)
+1. [SPA Mode](https://github.com/sveltejs/sapper/issues/383)
+
+It would facilitate easier deployment on some new platforms:
+
+1. [Azure Functions](https://azure.microsoft.com/en-gb/services/functions)
+1. [AWS Lambda](https://aws.amazon.com/lambda/)
+1. [Google Cloud Functions](https://cloud.google.com/functions)
+1. [IBM Cloud Functions](https://www.ibm.com/uk-en/cloud/functions)
+1. [IPFS](https://ipfs.io/)
+1. [Begin.com](https://begin.com/)
+1. [Netlify](https://www.netlify.com/)
+1. [Cloudflare Workers](https://blog.cloudflare.com/introducing-cloudflare-workers/)
+1. [Fastly Edge](https://www.fastly.com/products/edge-compute)
+
+## Detailed design
+
+> This is the bulk of the RFC.
+
+> Explain the design in enough detail for somebody
+familiar with the framework to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+* Client-side bundle goes into static dir (SPA)
```

Comment:
When you say "static dir" I'm guessing you mean `__sapper__/build/client/` and not `static/`?

Reply:
That is my understanding.

There is additional complexity here. We obviously need this information in the server builder as well, but if we are moving towards facilitating 'faas' deployment then we might want to consider greater support for plopping your assets on a CDN, there have been some issues about this in the past. I think this is quite difficult right now, if it is possible at all.

Reply:
The main difficulty right now that I'm aware of is that the images don't have hashes in their names, so they're not cacheable. I sent a PR to the template to address this: https://github.com/sveltejs/sapper-template/pull/248

Reply:
Re "if it is possible at all": Right now I'm deploying my sapper app to a CDN by parsing the routes directly out of manifest source code and generating CDN configuration for it.